### PR TITLE
Zstandard decoder implementation #1992

### LINKED ISSUE #1992: [v2] Add support for Content-Encoding: zstd
See: https://tools.ietf.org/html/rfc8478

### Minimum requirements

:moneybag: **You can get paid to complete this issue! [Please read the docs for more information](https://urllib3.readthedocs.io/en/latest/contributing.html#getting-paid-for-your-contributions).**

- [ ] Evaluate all zstandard packages that are available for Python, which makes sense for us to support? One, multiple?
- [ ] Add support for a `ZstandardDecoder`, don't enable `zstd` in `Accept-Encoding` by default though.
- [ ] Add documentation for the feature
- [ ] Add test cases for the feature, see `BrotliDecoder` and `Content-Encoding: brotli` tests for inspiration.

### PR DESCRIPTION
**Pull request by mamico**
I've evaluated https://github.com/animalize/pyzstd and https://github.com/indygreg/python-zstandard, they are really close in terms of features and activities.

I initially implemented for both libraries, but chose pyzstd because I was unable to raise the exception with python-zstandard in streaming mode (conform to the test https://github.com/urllib3/urllib3/compare/main...mamico:zstd?expand=1#diff-f853b537efd0c07ee4e2affffe963ac50d718f5894aa4e6f4e57c5f98ad98599R283).

Closes #1992

### CODE REVIEW

**Comment by sethmlarson** [comment]
Thanks for opening this! That's intriguing that you couldn't raise an error during streaming mode for `python-zstandard`. Of the [three projects listed on the Zstandard page](https://facebook.github.io/zstd/#other-languages) I'd been hoping we could make that one work. Is this something the maintainers of that project are aware of?

**Comment by sethmlarson** [comment]
Also @mamico if you're not aware that issue is [eligible to be paid after completion](https://urllib3.readthedocs.io/en/latest/contributing.html#getting-paid-for-your-contributions). Are you interested in pursuing that? If so let us know and we can help with the rest of the process too!

**Comment by mamico** [comment]
> Thanks for opening this! That's intriguing that you couldn't raise an error during streaming mode for `python-zstandard`. Of the [three projects listed on the Zstandard page](https://facebook.github.io/zstd/#other-languages) I'd been hoping we could make that one work. Is this something the maintainers of that project are aware of?

That's my problem:

```
>>> import zstandard
>>> ctx = zstandard.ZstdDecompressor()
>>> zfoo = zstandard.compress(b"foo")
```

decompressing all the data is fine:

```
>>> obj = ctx.decompressobj()
>>> obj.decompress(zfoo)
b'foo'
```

decompressing in streaming mode is fine too

```
>>> obj = ctx.decompressobj()
>>> b''.join([obj.decompress(good[i:i+1]) for i in range(len(zfoo))])
b'foo'
```

But, if I send something wrong (short), I am not able to know that there is a problem.

```
>>> obj = ctx.decompressobj()
>>> obj.decompress(b'bar')
b''
```

It is different if the data is enough

```
>>> obj = ctx.decompressobj()
>>> obj.decompress(b'fooooooooooooo')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
zstd.ZstdError: zstd decompressor error: Unknown frame descriptor
```

Instead, `pyzstd` provides the `at_frame_edge` and` needs_input` flags which help to recognize problems.

If you think that is ok to change the test and raise exception only when there is enough data, it's easy for me to change the library. 

We could also leave the current implementation as a fallback: so that, for example, if you installed `pyzstd` it still works, but `python-zstandard` is the supported library.

**Comment by sethmlarson** [comment]
@mamico Looked at the issue tracker for python-zstandard, would this feature request be the one needed to implement this? https://github.com/indygreg/python-zstandard/issues/161

File: src/urllib3/response.py
Comment (sethmlarson) [inline src/urllib3/response.py:163]: This line seems more complex than it needs to be, maybe we can extract the empty case into it's own block for better readability?

**Review by sethmlarson — COMMENTED** [review]
This PR looks great content-wise, here's a handful of comments for you to get this one across the finish line once when we decide on the python-zstandard issue:

**Comment by pquentin** [comment]
> @mamico Looked at the issue tracker for python-zstandard, would this feature request be the one needed to implement this? [indygreg/python-zstandard#161](https://github.com/indygreg/python-zstandard/issues/161)

Given the thumbs up on that comment I thought that it was the issue, so I mentioned it to @indygreg. He kindly implemented this feature, apparently spending multiple hours on the project to help adoption here. I tried his fix and it did not help. Looking more closely, I don't think `unused_data` is what we're looking for:

```
>>> import zstandard as zstd
>>> zstd.__version__
'0.18.0.dev0'
>>> obj = zstd.ZstdDecompressor().decompressobj()
>>> obj.decompress(b"foo")
b''
>>> obj.unused_data
b''
```

@indygreg Please accept my apologies for requesting work that was not ultimately useful to us. :(

To make sure that decompressing b"foo" fails like in the other backends, we can address this in two places:

 * python-zstandard: ask to raise an exception for this case, like all other backends do (adding the same attributes as pyzstd seems like an inferior solution to me)
 * urllib3: argue that the test is not realistic, and only test a longer string like `b"foooooooooooooo"` is fine

To avoid requesting more work to @indygreg (and because there must be a reason why the zstd libraries don't raise in this case), I think using a longer test string is fine, achieves the same result in practice, and allows us to use the better Python Zstandard library.

**Comment by mamico** [comment]
I've changed the library used from `pyzstd` to `python-zstandard` and made changes according to the review (thank for detailed work).

File: src/urllib3/response.py
Comment (pquentin) [inline src/urllib3/response.py:166]: For consistency with other decoders, can we please use an early return and check for all falsy values?

```suggestion
            if not data:
                return data

            return self._obj.decompress(data)  # type: ignore[no-any-return]
```

File: test/__init__.py
Comment (pquentin) [inline test/__init__.py:156]: Unlike brotlipy, we only support one Zstandard library:

```suggestion
        zstd is None, reason="only run if python-zstandard is installed"
```

File: test/__init__.py
Comment (pquentin) [inline test/__init__.py:162]: Unlike brotlipy, we only support one Zstandard library:

```suggestion
        zstd is None, reason="only run if python-zstandard is installed"
```

File: test/test_response.py
Comment (pquentin) [inline test/test_response.py:288]: Can we use pytest.mark.paramettrize with marks as done in `test_make_headers` to avoid duplicating those tests a fourth time? (Could be a follow-up PR if you prefer.)

**Review by pquentin — CHANGES_REQUESTED** [review]
Thanks! This looks good. I left a few comment to make sure [the contribution is polished](https://sage.thesharps.us/2014/09/01/the-gentle-art-of-patch-review/).

File: test/test_response.py
Comment (mamico) [reply test/test_response.py]: Ok, for a follow-up.

File: docs/advanced-usage.rst
Comment (sethmlarson) [reply docs/advanced-usage.rst:548]: Brotli uses its Content-Encoding value `br` so we could do `zstd` to continue that tradition.

**Comment by indygreg** [comment]
I'm unsure how much work it would be, but if you care about achieving peak performance, I would recommend not using `zstandard`'s stdlib compatible API for decompression (`dctx.decompressionobj()`). Instead, something like the _stream reader_ interface (https://python-zstandard.readthedocs.io/en/latest/decompressor.html#zstddecompressionreader) would be better.

The reason is because the stdlib API requires excessive memory allocations and copies. Since the `decompress()` API doesn't know the final output size, it can't allocate a fixed output buffer for the return value. Instead, it allocates several buffers and concatenates them to return a final value. With APIs like `read(n)`, you know how many bytes are requested and can decompress directly into the buffer that is ultimately returned. For pathological cases, the overhead of the excessive buffer allocations and copies with the stdlib API can be noticeable.

Modern versions of Python do define an optional `max_length` argument to `decompress()` to address this issue. But I don't have it implemented in `zstandard` yet and I suspect a lot of software in the wild doesn't use it in order to maintain compatibility with older Python versions while keeping source code simpler. The `io.RawIOBase` interface is a better model for applying filtering on a stream anyway.

Unfortunately, it looks like `urllib3`'s `ContentDecoder` interface uses the inefficient stdlib decompression API design. So porting away is probably non-trivial. But I thought I'd make you aware of this known performance issue. It's really too bad the compression APIs in the standard library are poorly designed from a performance perspective.

**Comment by mamico** [comment]
> I'm unsure how much work it would be, but if you care about achieving peak performance, I would recommend not using `zstandard`'s stdlib compatible API for decompression (`dctx.decompressionobj()`). 

@indygreg what do you think about https://github.com/indygreg/python-zstandard/pull/141 proposed by @animalize (the author of pyzstd)?  Can’t that be the solution to use a simple API with good performance?

**Comment by animalize** [comment]
> But, if I send something wrong (short), I am not able to know that there is a problem.

In `zstandard`, the same problem is that when the data is incomplete, there is no way to know.

Another problem is when there are mutiple frames, only the first is decompressed.
But this can be supported in the future.

> The reason is because the stdlib API requires excessive memory allocations and copies.

`zstandard` is using `.write()` with `"y#"` format, it generates a `bytes` intermediate object (extra copy).

In `pyzstd`, `compress_stream()` and `decompress_stream()` use a `memoryview` to avoid extra copy when writing. You may improve this.

> Since the decompress() API doesn't know the final output size, it can't allocate a fixed output buffer for the return value. Instead, it allocates several buffers and concatenates them to return a final value.

If the decompressed size is unknown, and user needs to get the whole decompressed data as a `bytes` object, they will also implement such function themselves, but it may be less efficient and error-prone.

File: test/__init__.py
Comment (pquentin) [inline test/__init__.py:154]: Why change `only` to `has`? `only` was fine, and consistent with other fixtures like `onlySecureTransport`

Comment (mamico) [reply test/__init__.py]: Changed back to `only`.

**Comment by animalize** [comment]
> In zstandard, the same problem is that when the data is incomplete, there is no way to know.

This statement is wrong, it has a `.eof` flag.
You can use it to raise exception:
```python
        def flush(self) -> bytes:
            ret = self._obj.flush()
            if not self._obj.eof:
                raise Exception()
            return ret
```

**Comment by pquentin** [comment]
Good find @animalize!

@indygreg OK now we have a good reason to ask for a release, I confirmed that the code above lets us raise correctly with `b"foo"` as input (which after all seemed important to us as bad settings and short responses are common on the web). The `eof` attribute was only added when you worked on `unused_data` though: https://github.com/indygreg/python-zstandard/commit/1dc39cfea821893268176dae754355ec19609ce0.

**Comment by pquentin** [comment]
Thanks! Can you please upload a source distribution too to fix the PyPy and CPython 3.11 builds? (They were working with 0.17.0)

File: src/urllib3/response.py
Comment (animalize) [inline src/urllib3/response.py:170]: Strictly speaking, the message can be this: zstd data ends in an incomplete frame.

**Review by sethmlarson — COMMENTED** [review]
One thought: since we can't guarantee correctness using zstandard<0.17 we should update our version ranges for zstandard in the `zstd` extra to be `>=0.18` and check `zstandard.__version__` as a part of detecting whether we expose `ZstdDecoder`. Add this to the documentation as well.

**Comment by indygreg** [comment]
> Thanks! Can you please upload a source distribution too to fix the PyPy and CPython 3.11 builds? (They were working with 0.17.0)

0.18.0 sdist uploaded.

File: src/urllib3/response.py
Comment (animalize) [reply src/urllib3/response.py:170]: It's a little long, how about "zstd data is incomplete"?

**Comment by animalize** [comment]
Should `DeflateDecoder` / `GzipDecoder` / `BrotliDecoder` check `.eof` flag as well?
In current code, they don't check it.

**Comment by sethmlarson** [comment]
@animalize I'm not sure! Maybe those decoders not checking for `eof` and us receiving no reports of issues is a sign users aren't receiving truncated streams? Our usage of `eof` in ZstdDecoder is to work-around the [issue you've raised](https://github.com/facebook/zstd/issues/3169) but I don't think we're as worried about partial streams there as we are about any "amount of data that's definitely not zstandard" not raising an error.

**Comment by sethmlarson** [comment]
@mamico Went ahead and push my review comments since they were very small and the only outstanding work to get this one past the finish line. Let me know if you have concerns here.

File: docs/advanced-usage.rst
Comment (pquentin) [inline docs/advanced-usage.rst:547]: Realized thanks to the note below that we should call this package `zstandard` after its name on PyPI.

```suggestion
and is supported by urllib3 if the `zstandard <https://pypi.org/project/zstandard/>`_ package is installed.
```

image

File: docs/advanced-usage.rst
Comment (pquentin) [inline docs/advanced-usage.rst:567]: Let's avoid a redirect:

```suggestion
        "https://www.facebook.com/",
```

File: docs/reference/urllib3.response.rst
Comment (pquentin) [inline docs/reference/urllib3.response.rst:33]: I think having ZstdDecoder before MultiDecoder would be nicer.

image

**Review by pquentin — APPROVED** [review]
I left a few nits, please merge after you've considered them.

File: docs/advanced-usage.rst
Comment (sethmlarson) [inline docs/advanced-usage.rst:547]: Sphinx doesn't like having two links with the same text, maybe we include the word "package" within the hyperlink text here?

**Comment by sethmlarson** [comment]
The failure on macOS+Python 3.10 looks unrelated, an exit code of `-14` but not during a test case that changed. Curious. I don't think we need to block this PR on that failure.

**Review by sethmlarson — APPROVED** [review]
This looks wonderful, thanks to everyone who contributed to this PR! @mamico @animalize @indygreg :bow:

**Pull request merged**
